experiments/modern/src/utils.js

```diff
@@ -139,3 +141,56 @@ export function findDescendantByTypeAndId(node, type, id) {
 
   return null;
 }
+
+function findDirectDescendantById(node, id) {
+  return node.children.find(item => item.xmlNode.attributes.id === id);
+}
+
+// Search up the tree for <Elements> nodes that are in node's lexical scope.
+// return the first child of an <Elements> that matches id
+export function findElementById(node, id) {
+  let currentNode = node;
+  while (currentNode.parent) {
+    let parent = currentNode.parent;
+    const children = parent.children;
+    for (let i = 0; i < children.length; i++) {
+      if (children[i] === node) {
+        break;
+      }
+
+      if (children[i] instanceof Elements) {
+        const element = findDirectDescendantById(children[i], id);
+        if (element) {
+          return element;
+        }
+      }
+    }
+    currentNode = parent;
+  }
+
+  return null;
+}
+
+// Search up the tree for a <GroupDef> node that is in node's lexical scope and matches id.
+export function findGroupDefById(node, id) {
+  let currentNode = node;
+  while (currentNode.parent) {
+    let parent = currentNode.parent;
+    const children = parent.children;
+    for (let i = 0; i < children.length; i++) {
+      if (children[i] === node) {
+        break;
+      }
+
+      if (
+        children[i] instanceof GroupDef &&
+        children[i].xmlNode.attributes.id === id
+      ) {
+        return children[i];
+      }
+    }
+    currentNode = parent;
+  }
+
+  return null;
```

Comment:
Could we find a way to share logic between these two? Maybe write a generic “find in lexical scope” function which takes a predicate callback?

Reply:
yea, sounds great!